data-prepper-plugins/date-processor/src/main/java/com/amazon/dataprepper/plugins/processor/date/DateProcessor.java

```diff
@@ -12,37 +12,164 @@
+                    zonedDateTime = getZonedDateTime(sourceTimestamp, formatter);
+                    if (zonedDateTime != null)
+                        break;
+                }
+                if (zonedDateTime == null) {
+                    LOG.debug("Unable to parse {} with provided patterns. {}", sourceTimestamp, parsingException.getMessage());
+                }
+            }
+            if (zonedDateTime != null)
+                record.getData().put(dateProcessorConfig.getDestination(), zonedDateTime);
+        }
+        return records;
+    }
+
+    private ZoneId buildZoneId(final String timezone) {
+        try {
+            return ZoneId.of(timezone);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Unsupported timezone provided.");
+        }
+    }
+
+    private Locale parseLocale(final String localeString) {
+        boolean isBCP47Format = localeString.contains("-");
+
+        final String[] localeFields;
+        if (isBCP47Format) {
+            localeFields = localeString.split("-");
+        }
+        else
+            localeFields = localeString.split("_");
+
+        switch (localeFields.length) {
+            case 1:
+                return new Locale(localeFields[0]);
+            case 2:
+                return new Locale(localeFields[0], localeFields[1]);
+            case 3:
+                return new Locale(localeFields[0], localeFields[1], localeFields[2]);
+            default:
+                throw new IllegalArgumentException("Invalid locale format. Only language, country and variant are supported.");
+        }
+    }
+
+    private void extractKeyAndFormatters(final Map<String, List<String>> match) {
+        for (final Map.Entry<String, List<String>> entry : match.entrySet()) {
+            keyToParse = entry.getKey();
+            dateTimeFormatters = (entry.getValue().stream().map(this::getSourceFormatter).collect(Collectors.toList()));
+        }
+        if (dateTimeFormatters.isEmpty())
+            throw new IllegalArgumentException("At least 1 pattern is required.");
+    }
+
+    private DateTimeFormatter getSourceFormatter(final String pattern) {
+        final ZonedDateTime zonedDateTimeForDefaultValues = ZonedDateTime.now();
+        return new DateTimeFormatterBuilder()
+                .appendPattern(pattern)
+                .parseDefaulting(ChronoField.YEAR, zonedDateTimeForDefaultValues.getYear())
+                .parseDefaulting(ChronoField.MONTH_OF_YEAR, zonedDateTimeForDefaultValues.getMonthValue())
+                .parseDefaulting(ChronoField.DAY_OF_MONTH, zonedDateTimeForDefaultValues.getDayOfMonth())
+                .parseDefaulting(ChronoField.HOUR_OF_DAY, zonedDateTimeForDefaultValues.getHour())
+                .parseDefaulting(ChronoField.MINUTE_OF_HOUR, zonedDateTimeForDefaultValues.getMinute())
+                .parseDefaulting(ChronoField.SECOND_OF_MINUTE, zonedDateTimeForDefaultValues.getSecond())
```

Comment:
I want to point out that I used default values for year, month, day, hour, minute, seconds if it's not present in the value the customer wants to parse. Let me know if there are any other default values we should use. 
I don't think we should log parse failure since most logs won't have all the fields to build date time object.